web/packages/teleterm/src/ui/services/modals/modalsService.ts

```diff
@@ -58,20 +67,44 @@ export class ModalsService extends ImmutableStore<State> {
    * old dialog with the new one.
    * The old dialog is canceled, if possible.
    *
-   * The returned closeDialog function can be used to close the dialog and automatically call the
-   * dialog's onCancel callback (if present).
+   * The passed `abortSignal` or the returned `closeDialog` function can be used to close the dialog
+   * and automatically call the dialog's onCancel callback (if present).
    */
-  openRegularDialog(dialog: Dialog): { closeDialog: () => void } {
+  openRegularDialog(
+    dialog: Dialog,
+    abortSignal?: AbortSignal
+  ): {
+    closeDialog: () => void;
+  } {
+    const onCancelDialog = () => dialog['onCancel']?.();
+    const sharedSignal = this.getSharedAbortSignal(abortSignal);
+    if (sharedSignal.aborted) {
+      onCancelDialog();
+      return {
+        closeDialog: () => {},
+      };
+    }
+
     this.state.regular?.['onCancel']?.();
+
+    const closeDialog = () => {
+      sharedSignal.removeEventListener('abort', closeDialog);
+      // Do we still have the same dialog open?
+      if (dialog !== this.state.regular) {
+        return;
+      }
```

Comment:
What scenario does this protect against?

Reply:
I was thinking of a following case (everything happens in the same scope so that the abort signal is not garbage collected):
1. Open a dialog A with abort signal.
2. Close it through some interaction.
3. Open dialog B without a signal.
4. Abort the signal for A. 
5. B is closed.

Since we didn't remove the listener for the dialog A we would close the currently open dialog (`closeRegularDialog()` doesn't have access to the signal and can't unregister the abort listener).
I will add a test case for it.

Reply:
Instead of handling this case separately, what do you think about guaranteeing that when a regular dialog gets canceled, we always remove the listener for it? This way no matter how we interact with the dialog, we can be sure that the listener is removed, instead of having to worry about multiple sets of interactions which may or may not remove the listener.

Reply:
Yeah, that would be the best. The only problem is that the function that the `closeRegularDialog` can't access the `closeDialog` function from `open*Dialog` which has access to the signal. I think the solution is to keep the function that closes the dialog along with `dialog` in the state. This will allow us to always remove the listener.

Reply:
Done.